code/window_snap.py

```diff
@@ -0,0 +1,221 @@
+"""Tools for voice-driven window management.
+
+Originally from dweil/talon_community - modified for newapi by jcaw.
+
+"""
+
+# TODO: Map keyboard shortcuts to this manager once Talon has key hooks on all
+#   platforms
+
+import time
+from operator import xor
+from typing import Optional
+
+from talon import ui, Module, Context, actions
+
+
+def sorted_screens():
+    """Return screens sorted by their topmost, then leftmost, edge.
+
+    Screens will be sorted left-to-right, then top-to-bottom as a tiebreak.
+
+    """
+
+    return sorted(
+        sorted(ui.screens(), key=lambda screen: screen.visible_rect.top),
+        key=lambda screen: screen.visible_rect.left,
+    )
+
+
+def _set_window_pos(window, x, y, width, height):
+    """Helper to set the window position."""
+    # TODO: Special case for full screen move - use os-native maximize, rather
+    #   than setting the position?
+    rect = ui.Rect(round(x), round(y), round(width), round(height))
+    window.rect = rect
+    # HACK: If on Windows and the window was maximised, this can leave the
+    #   window in a weird semi-maximised state. Some windows will be tiny or
+    #   invisible in this state. Repeating seems to solve this.
+    actions.sleep("250ms")
+    window.rect = rect
+
+
+def _bring_forward(window):
+    current_window = ui.active_window()
+    try:
+        window.focus()
+        current_window.focus()
```

Comment:
Feels a little weird to maintain the focus of the current window to me. What's the reasoning for this, out of curiosity?

Reply:
Copying original behaviour - just improving the feel slightly. Use case is presumably something like, you're in the program you want and you want to put, say, Firefox on the left for reference:

`"snap left, snap firefox left"` vs. `"snap firefox left, snap emacs right"`

First is more concise, but the second seems more intuitive. You may be right, forcing focus for the moved window seems a little better.

Reply:
I hadnt known this behavior was available, but I like the idea. I assumed first time I read it that it would maintain focus on whatever you are in. It could be an option to preserve or not.

I have a similar situation with a global command "shell rerun". It reruns whatever was in the shell history without losing focus on current app. Was thinking of something like "shell rerun stay" to keep focus on shell after. Many times though I just want to rerun the shell ad watch while keeping focus in my editor.

Reply:
Yeah, maybe consider adding a setting for this behavior before merging? i.e. while we wait for the official fix.

It really feels weird to me, so I'd probably vote for a default of changing focus to the app that is snapped in settings.talon? I think it'll be really confusing to newbies.

Reasonable compromise?

Reply:
I can add a setting. How are settings managed in knausj at the moment?

Reply:
Oops, missed this. the setting's declared/defaulted and used as appropriate .py, in this case switcher.py.

Then the user can set it here as desired.
https://github.com/knausj85/knausj_talon/blob/master/settings.talon 

Goal being one file to branch for most settings.

Reply:
This can certainly wait for another day too, honestly